File: common/src/main/java/org/xpilot/common/Rules.java
package org.xpilot.common;/*
 * XPilot NG, a multiplayer space war game.
 *
 * Copyright (C) 1991-2001 by
 *
 *      Bjørn Stabell        <bjoern@xpilot.org>
 *      Ken Ronny Schouten   <ken@xpilot.org>
 *      Bert Gijsbers        <bert@xpilot.org>
 *      Dick Balaska         <dick@xpilot.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

public class Rules{

/*
 * Bitfield definitions for playing mode.
 */
public static final int CRASH_WITH_PLAYER = (1<<0);
public static final int BOUNCE_WITH_PLAYER = (1<<1);
public static final int PLAYER_KILLINGS = (1<<2);
public static final int LIMITED_LIVES = (1<<3);
public static final int TIMING = (1<<4);
public static final int PLAYER_SHIELDING = (1<<6);
public static final int LIMITED_VISIBILITY = (1<<7);
public static final int TEAM_PLAY = (1<<8);
public static final int WRAP_PLAY = (1<<9);
public static final int ALLOW_NUKES = (1<<10);
public static final int ALLOW_CLUSTERS = (1<<11);
public static final int ALLOW_MODIFIERS = (1<<12);
public static final int ALLOW_LASER_MODIFIERS = (1<<13);
public static final int ALLIANCES = (1<<14);

/*
 * Client uses only a subset of them:
 */
public static final int CLIENT_RULES_MASK = (WRAP_PLAY|TEAM_PLAY|TIMING|LIMITED_LIVES|ALLIANCES);
/*
 * Old player status bits, currently only used in network protocol.
 * The bits that the client needs must fit into a byte,
 * so the first 8 bitvalues are reserved for that purpose.
 */
public static final int OLD_PLAYING = (1L<<0)	;	/* alive or killed */
public static final int OLD_PAUSE = (1L<<1); 	/* paused */
public static final int OLD_GAME_OVER = (1L<<2);		/* waiting or dead */

    int		lives;
    long	mode;

}

File: KXPilot/shared/src/commonMain/kotlin/org/lambertland/kxpilot/common/Rules.kt
package org.lambertland.kxpilot.common

// ---------------------------------------------------------------------------
// Ported from: common/rules.h
// ---------------------------------------------------------------------------

/**
 * Game-mode bit flags.  The [Rules.mode] field is a bitfield of these values.
 * Maps to the `#define` constants in common/rules.h.
 */
object GameMode {
    val CRASH_WITH_PLAYER: Long = 1L shl 0
    val BOUNCE_WITH_PLAYER: Long = 1L shl 1
    val PLAYER_KILLINGS: Long = 1L shl 2
    val LIMITED_LIVES: Long = 1L shl 3
    val TIMING: Long = 1L shl 4
    val PLAYER_SHIELDING: Long = 1L shl 6
    val LIMITED_VISIBILITY: Long = 1L shl 7
    val TEAM_PLAY: Long = 1L shl 8
    val WRAP_PLAY: Long = 1L shl 9
    val ALLOW_NUKES: Long = 1L shl 10
    val ALLOW_CLUSTERS: Long = 1L shl 11
    val ALLOW_MODIFIERS: Long = 1L shl 12
    val ALLOW_LASER_MODIFIERS: Long = 1L shl 13
    val ALLIANCES: Long = 1L shl 14

    /** Subset of mode flags the client cares about. */
    val CLIENT_RULES_MASK: Long =
        WRAP_PLAY or TEAM_PLAY or TIMING or LIMITED_LIVES or ALLIANCES
}

/**
 * Legacy player-status bits used only in the network protocol.
 * Maps to `OLD_PLAYING`, `OLD_PAUSE`, `OLD_GAME_OVER` in common/rules.h.
 */
object OldPlayerStatus {
    val OLD_PLAYING: UInt = 1u shl 0
    val OLD_PAUSE: UInt = 1u shl 1
    val OLD_GAME_OVER: UInt = 1u shl 2
}

/**
 * Server game configuration (lives + mode).
 * Maps to C `rules_t { int lives; long mode; }` in common/rules.h.
 */
data class Rules(
    val lives: Int,
    val mode: Long,
)
